app/src/main/java/com/xmartlabs/template/controller/EntityController.java

```diff
@@ -0,0 +1,145 @@
+package com.xmartlabs.template.controller;
+
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+
+import com.annimon.stream.Optional;
+import com.raizlabs.android.dbflow.sql.language.property.Property;
+import com.xmartlabs.template.helper.function.BiFunction;
+import com.xmartlabs.template.helper.function.Function;
+import com.xmartlabs.template.model.EntityWithId;
+
+import java.util.List;
+
+import io.reactivex.Completable;
+import io.reactivex.Flowable;
+import io.reactivex.Observable;
+import io.reactivex.Single;
+import io.reactivex.schedulers.Schedulers;
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public abstract class EntityController<Id, D extends EntityWithId<Id>, C> extends Controller {
+  private final EntityDao<Id, D, C> entityDao;
+  private final EntityServiceProvider<Id, D> entityServiceProvider;
+
+  @CheckResult
+  @NonNull
+  @SuppressWarnings({"unused", "WeakerAccess", "unchecked"})
+  protected Flowable<List<D>> getEntities(@NonNull Single<List<D>> serviceCall, @NonNull C... conditions) {
+    return Flowable
+        .concatArrayDelayError(
+            entityDao.getEntities(conditions).toFlowable(),
+            serviceCall
+                .compose(entityServiceProvider.applySingleServiceTransformation())
+                .toFlowable()
+        )
+        .subscribeOn(Schedulers.io())
+        .observeOn(Schedulers.io())
+        .scan((databaseEntities, serviceEntities) ->
+            entityDao.deleteAndInsertEntities(serviceEntities, conditions).blockingGet()
+        )
+        .compose(applyFlowableIoSchedulers());
+  }
+
+  @CheckResult
+  @NonNull
+  @SuppressWarnings({"unused", "WeakerAccess"})
+  protected Observable<D> getEntity(@NonNull Function<Id, Single<D>> serviceCall, @NonNull Id id) {
+    return Observable.mergeDelayError(
+        entityDao.getEntity(id).toObservable(),
+        getServiceEntity(serviceCall, id).toObservable()
+    );
+  }
+
+  @CheckResult
+  @NonNull
+  @SuppressWarnings({"unused", "WeakerAccess"})
+  protected Single<D> getEntityFromDbIfPresentOrGetFromService(@NonNull Function<Id, Single<D>> serviceCall,
+                                                               @NonNull Id id) {
+    return entityDao.getEntity(id)
+        .switchIfEmpty(getServiceEntity(serviceCall, id).toMaybe())
```

Comment:
Why? I think is something like [this](https://github.com/xmartlabs/Android-Base-Project/pull/42/files/b07ccc17eff7bc3e5230ee9b5394fe7ef521a648#diff-a421c547b6e523743265959721a54cb0R32)